### ISSUE TITLE
iter(classmethod, sentinel) broken for Argument Clinic class methods?

### ISSUE THREAD

**Issue by 652782da-b5f3-46b2-ae1b-1f73662bb759** [python/cpython#74709]
BPO | [30524](https://bugs.python.org/issue30524)
--- | :---
Nosy | @mjpieters, @vstinner, @serhiy-storchaka
PRs | <li>python/cpython#1886</li><li>python/cpython#2022</li><li>python/cpython#2030</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-06-09.20:45:17.849>
created_at = <Date 2017-05-31.12:01:15.958>
labels = ['interpreter-core', 'type-bug']
title = 'iter(classmethod, sentinel) broken for Argument Clinic class methods?'
updated_at = <Date 2017-06-20.20:34:33.131>
user = 'https://github.com/mjpieters'
```

bugs.python.org fields:
```python
activity = <Date 2017-06-20.20:34:33.131>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-06-09.20:45:17.849>
closer = 'vstinner'
components = ['Interpreter Core']
creation = <Date 2017-05-31.12:01:15.958>
creator = 'mjpieters'
dependencies = []
files = []
hgrepos = []
issue_num = 30524
keywords = ['3.6regression']
message_count = 10.0
messages = ['294835', '294836', '294841', '294845', '294847', '295515', '295535', '295566', '295572', '296486']
nosy_count = 3.0
nosy_names = ['mjpieters', 'vstinner', 'serhiy.storchaka']
pr_nums = ['1886', '2022', '2030']
priority = 'high'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue30524'
versions = ['Python 3.6']
```

</p></details>

**Comment by 652782da-b5f3-46b2-ae1b-1f73662bb759** [comment]
I'm not sure where exactly the error lies, but bpo-27128 broke iter() for Argument Clinic class methods. The following works in Python 3.5, but not in Python 3.6:

```py
from datetime import datetime
from asyncio import Task

next(iter(datetime.now, None))
next(iter(Task.all_tasks, None))
```


In 3.6 StopIteration is raised:

```pycon
>>> next(iter(datetime.now, None))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
>>> next(iter(Task.all_tasks, None))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```


(In 3.5 a `datetime.datetime` and `set` object are produced, respectively)

The only thing these two methods have in common is that they are class methods with no arguments, parsed out by the Argument Clinic generated code (so using \_PyArg_Parser).

What appears to have changed is that iter() was switched from using PyObject_Call to \_PyObject_FastCall, see https://github.com/python/cpython/commit/99ee9c70a73ec2f3db68785821a9f2867c3f637f

**Comment by 652782da-b5f3-46b2-ae1b-1f73662bb759** [comment]
Forgot to addthis: this bug was found via https://stackoverflow.com/questions/44283540/iter-not-working-with-datetime-now

**Comment by serhiy-storchaka** [comment]
It works in 3.5 and 3.7, but raises a StopIteration in 3.6.

**Comment by vstinner** [comment]
Oh, it's my fault: it's a bug coming from FASTCALL optimizations. The strange thing is that the bug wasn't catched by the giant Python test suite :-(

I knew that \_PyStack_UnpackDict() has a bug in Python 3.6, but I completely forgot to fix it :-(

https://github.com/python/cpython/pull/1886 fixes the bug, but has not test yet.

**Comment by serhiy-storchaka** [comment]
next(iter(datetime.now, None)) can be turned into a nice test.

**Comment by vstinner** [comment]
New changeset f0ff849adc6b4a01f9d1f08d9ad0f1511ff84541 by Victor Stinner in branch '3.6':
bpo-30524: Fix \_PyStack_UnpackDict() (bpo-1886)
https://github.com/python/cpython/commit/f0ff849adc6b4a01f9d1f08d9ad0f1511ff84541

**Comment by vstinner** [comment]
New changeset 3b5cf85edc188345668f987c824a2acb338a7816 by Victor Stinner in branch 'master':
bpo-30524: Write unit tests for FASTCALL (bpo-2022)
https://github.com/python/cpython/commit/3b5cf85edc188345668f987c824a2acb338a7816

**Comment by vstinner** [comment]
New changeset b7577456c430283f8b7ec4e914b701cb943cc69b by Victor Stinner in branch '3.6':
bpo-30524: Write unit tests for FASTCALL (bpo-2022) (bpo-2030)
https://github.com/python/cpython/commit/b7577456c430283f8b7ec4e914b701cb943cc69b

**Comment by vstinner** [comment]
Sorry for the regression, sadly, it wasn't catch before by any test. I added a lot of new tests, so we should cover more cases. Oh, and the bug has been fixed in 3.6 :-)

**Comment by vstinner** [comment]
bpo-30473 has been marked as a duplicate of this bug.
